### ISSUE TITLE
Open items w.r.t. resource based processing

### ISSUE THREAD

**Issue by matz-e** [matz-e/lobster#279]
*This is a sort-of meta-issue.*

This also shouldn't change anything if `n` cores = 1 task, where `n` is constant.

Currently, we don't really take multi-core tasks into account apart from setting the resource requirements. What should happen:

* [ ] At task creation, enough tasks should be created to fill the cores, removing the current assumption 1 core = 1 task. Yet, the work to be processed by different workflows should behave as it does currently? I.e. creating more tasks for workflows with more work available.
* [ ] We want to limit workflows/groups of workflows (this is a currently undocumented feature…) to only consume a certain amount of resources (cores, tasks, bandwidth), tune-able by the user?
* [ ] We want to use all resources possible (at all cost?). I.e., if we find out that there is a 2-core slot while only 4-core tasks are being processed, we'd want to fill that ideally. This could mean that we want to keep at least `n` 2-core tasks queued, where `n > 0`.

**Comment by klannon** [comment]
This is right on, and I think your third point is the key.  In my mind, the best division of labor would be that Lobster should just make sure that WQ is fed with some minimum number of tasks per *category = workflow*  I often hear in the discussion that WMAgent tries to make sure it has enough jobs of each workflow waiting in the queues to make sure that Glidein makes enough pilots and that there are jobs waiting to grab pilots when they become available.  I think we want to do the same thing--make sure that we feed WQ enough tasks of each category that some minimum number are waiting.  Then it will be up to WQ to make sure that it sorts through these tasks and puts the maximum number of waiting ones that it can into the running state.  Does that make sense?  Presumably the key feature here is having WQ stats per category?  Would this let Lobster see how many tasks are running and waiting in each category?

I think if we (which of course means our intrepid Lobster and WQ experts with me cheering you on...) have the above, then setting limits on running tasks per workflow in Lobster is pretty straightforward:  If you have a limit of N on a particular workflow = category, then if you have >= N running, just stop submitting any waiting (i.e. set the number of desired waiting jobs to 0).  I guess the only extra piece of logic will be in the initial startup phase where you try to fill the full resources, but I don't think handling that special case will be too hard either.

For the moment, I think that keeping the default making tasks in proportion to who has the most work to do is fine.  The tweak is that if any workflow drops below some threshold in waiting tasks, as long as it's not done, you make new tasks for it to bring it back up to the threshold, just to make sure that WQ can fill all the resources.

Not to namedrop, but I'm just going to mention @btovar to make sure he's looped in, since I think he's already working on the WQ stats per category idea.

**Comment by matz-e** [comment]
I've been struggling a bit coming up with a good generic way of doing this.  What I have come up with so far:

1. Mangle the number of unused cores into a number of cores to assign tasks
   to
2. Find out how many tasks can be created by each category (and workflow).
   Multiply by the number of cores and sum up.  This is the total amount of
   cores to be filled.
3. For each category, the number of cores allotted is the fraction of cores
   the category can fill over the number of cores all categories could
   provide with work (from (2)).  If the category is limited to a certain
   number of tasks running, take that number (multiplied by cores per task)
   as the allotment.
4. Within each category, do the same again, using workflows instead of
   categories.
5. Create tasks from each workflow.

There will always be a minimum of 1 task created per workflow. I'm not sure we can actually do anything about the last point in the issue…

**Comment by matz-e** [comment]
For the record, this should be considered done once #283 is fixed.
